### PR TITLE
Add accordion with descriptions to Topic pages

### PR DESCRIPTION
**Pull request by gemmaleigh**
Add an accordion with descriptions module to enhance subsections - a heading, description and list of links - when JavaScript is enabled.

This module has been [added to the GOV.UK accessibility sandbox](http://govuk-accessibility-sandbox.herokuapp.com/patterns/accordion-with-descriptions/02/).

**Accordion with descriptions**

_What is it?_

The pattern is a list of headings each with a description and some hidden content. When you select the heading or the description the hidden content is revealed. There is also a button to open and close all the content.

The pattern allows users to see a summary of available content which can then be expanded if a user wants to read more.

**With JavaScript disabled:**

js-disabled

**With JavaScript enabled:**

agile delivery gov uk

agile delivery gov uk - open

### CODE REVIEW

**Comment by benilovj** [comment]
Slightly off-topic, but will [manual sections](https://www.gov.uk/guidance/the-highway-code/rules-for-cyclists-59-to-82) be updated to use this module at some point as well, or is that different?

image

**Comment by gemmaleigh** [comment]
Hi @benilovj, I'm not sure what the plan for the existing manuals format is. That's probably a question for @markhurrell and @aduggin.

File: app/views/content_items/service_manual_topic.html.erb [hunk 1]
```diff
@@ -1,6 +1,8 @@
 <%= content_for :page_class, "service-manual-topic" %>
 <%= content_for :title, @content_item.title %>
 
+<%= javascript_include_tag 'modules/accordion-with-descriptions.js' %>
```
Comment (thehenster) [inline app/views/content_items/service_manual_topic.html.erb]: We shouldn't need the .js extension here :scissors:

File: app/assets/javascripts/modules/accordion-with-descriptions.js [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+      }
+
+      // Set the aria controls for the open/close all button value for all content items
+      var $openCloseButton = $element.find('.js-open-close');
+
+      $openCloseButton.attr('aria-controls', ariaControlsValue);
+
+      // Hide the content
+      $element.find('.subsection-content')
+        .addClass('js-hidden');
```
Comment (thehenster) [inline app/assets/javascripts/modules/accordion-with-descriptions.js]: Could the below `closeSection()` be used here?

File: app/assets/javascripts/modules/accordion-with-descriptions.js [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+        return false;
+      });
+
+      $element.find('.subsection-button').on('click', function(e) {
+        toggleSection($(this).parent().parent().next());
+        toggleIcon($(this).parent().parent());
+        toggleState($(this));
+        setOpenCloseAllText();
+        return false;
+      });
```
Comment (thehenster) [inline app/assets/javascripts/modules/accordion-with-descriptions.js]: A wee one. Assigning `$(this)` to a variable and reusing will be a tiny bit more memory efficient, faster and more readable (for me at least). _shurg_

File: app/assets/javascripts/modules/accordion-with-descriptions.js [hunk 4]
```diff
@@ -0,0 +1,160 @@
...
+        // Find out if the number of is-opens == total number of sections
+        if (openSubsections === totalSubsections) {
+          $openOrCloseAllButton.text('Close all');
+        } else {
+          $openOrCloseAllButton.text('Open all');
+        }
+      }
+
+      function toggleSection($node) {
+        if ($($node).hasClass('js-hidden')) {
```
Comment (thehenster) [inline app/assets/javascripts/modules/accordion-with-descriptions.js]: Is `$node` already a jQuery object?

**Comment by AndrewVos** [comment]
lgtm. :ship: 

Didn't we already have jquery included though?

**Comment by gemmaleigh** [comment]
This PR has now been moved to the [service manual frontend](https://github.com/alphagov/service-manual-frontend/pull/5).

**Pull request closed without merging**